File: src/main/java/name/seguri/java/leetcode/_20/Solution.java
package name.seguri.java.leetcode._20;

import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
  public boolean isValid(String s) {
    Deque<Character> stack = new ArrayDeque<>();
    for (int i = 0; i < s.length(); i++) {
      char c = s.charAt(i);
      switch (c) {
        case '(', '[', '{' -> stack.push(c);
        case ')', ']', '}' -> {
          if (stack.isEmpty()) {
            return false;
          }
          char matching = stack.pop();
          if (c == ')' && matching != '(') {
            return false;
          }
          if (c == ']' && matching != '[') {
            return false;
          }
          if (c == '}' && matching != '{') {
            return false;
          }
        }
        default -> throw new IllegalArgumentException(c + ": unexpected value");
      }
    }
    return stack.isEmpty();
  }
}

File: src/main/kotlin/name/seguri/kotlin/leetcode/_20/Solution.kt
package name.seguri.kotlin.leetcode._20

import kotlin.collections.ArrayDeque

class Solution {
  fun isValid(s: String): Boolean {
    val stack = ArrayDeque<Char>()
    val bracketPairs = mapOf(')' to '(', ']' to '[', '}' to '{')

    for (c in s) {
      when (c) {
        in "([{" -> stack.addFirst(c)
        in ")]}" -> {
          if (stack.isEmpty() || stack.removeFirst() != bracketPairs[c]) {
            return false
          }
        }
        else -> error("$c: unexpected character")
      }
    }

    return stack.isEmpty()
  }
}
